### ISSUE TITLE
Nicer validation messages

### ISSUE THREAD

**Issue by kahwee** [balderdashy/waterline#148]
I notice the validation messages could be nicer.

```
[ { data: '12345678',
       message: 'Validation error: "12345678" Rule "len(9)" failed.',
       rule: 'len',
       args: [Object] } ]
```

Is there plans for this, I'll like to help out with this as an enhancement. Any thoughts?

**Comment by eabovsky** [comment]
Bump... another ex PHP YII user here.

I gotta say, that framework got the validation / model right. Seriously a huge bummer/time-suck on re-inventing wheels because sails / waterline doesn't have this implemented right.

**Comment by robgrimball** [comment]
Just to throw in further request to this effort.  There is also the case where the error is unknown to sails/waterline that would be nice to be better addressed.  For instance, in the sails-mongo adapter, if an attempt to add a duplicate key is encountered, it reports back an "UNKNOWN_E" error, and drops the error that Mongo reported into a "raw" field for the error.  This means that in order to find out pertinent information about the error (like what key was duplicate) so you can handle it in the front end(ie flag the element for that key to the user visually), you have to parse this raw structure which can (and does) change from Mongo version to version.  Likewise, since it is apparently a "raw" error, it's structure seems to get tossed about from sails-mongo version to version.

So I guess what I am asking is that waterline expand it's handling of errors to some frequently used ones like duplicate indexes, and make that a requirement of adapters of all flavors to at least attempt to provide information about those sorts of errors (duplicate index name, value attempted, etc).  But also that raw errors get some sort of TLC so they are a bit more easily handled somehow.  The latter I'm not sure how much can be done about it, though the former I imagine could be implemented given most data stores have unique indexing of some sort, and if they don't they would just never return that sort of error.

In a general sense though I do think some of the above suggestions for revamping the validation and handling of errors would be good.

**Comment by sailsbot** [comment]
Thanks for posting, @kahwee.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
